presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java

```diff
@@ -0,0 +1,121 @@
+import static com.facebook.presto.sql.planner.PlanAssert.semiJoinNode;
+import static com.facebook.presto.sql.planner.PlanAssert.symbol;
+import static com.facebook.presto.sql.planner.PlanAssert.tableScanNode;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+
+public class TestLogicalPlanner
+{
+    private final LocalQueryRunner queryRunner;
+
+    public TestLogicalPlanner()
+    {
+        this.queryRunner = new LocalQueryRunner(testSessionBuilder()
+                .setCatalog("local")
+                .setSchema("tiny")
+                .build());
+
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpchConnectorFactory(queryRunner.getNodeManager(), 1),
+                ImmutableMap.<String, String>of());
+    }
+
+    @Test
+    public void testPlanDsl()
+    {
+        assertPlan("SELECT orderkey FROM orders WHERE orderkey IN (1, 2, 3)",
+                node(OutputNode.class,
+                        node(FilterNode.class,
+                                tableScanNode("orders"))));
+
+        String simpleJoinQuery = "SELECT o.orderkey FROM orders o, lineitem l WHERE l.orderkey = o.orderkey";
+        assertPlan(simpleJoinQuery,
+                anyNode(
+                        anyNode(
+                                node(JoinNode.class,
+                                        anyNode(tableScanNode("orders")),
+                                        anyNode(tableScanNode("lineitem"))))));
+
+        assertPlan(simpleJoinQuery,
+                anyNodesTree(
+                        node(JoinNode.class,
+                                anyNodesTree(),
+                                anyNodesTree())));
+
+        assertPlan(simpleJoinQuery, anyNodesTree(node(TableScanNode.class)));
+
+        assertPlan("SELECT * FROM orders WHERE orderkey = (SELECT orderkey FROM lineitem ORDER BY orderkey LIMIT 1)",
+                anyNodesTree(
+                        joinNode(ImmutableList.of(aliasPair("H1", "H2"), aliasPair("H1", "H2"), aliasPair("X", "Y")),
+                                projectNode(
+                                        tableScanNode("orders").with(symbol("orderkey").as("X")))
+                                        .with(symbol("hash").as("H1")),
+                                projectNode(
+                                        node(EnforceSingleRowNode.class,
+                                                anyNodesTree(
+                                                        tableScanNode("lineitem")
+                                                                .with(symbol("orderkey").as("Y")))))
+                                        .with(symbol("hash").as("H2")))));
+
+        assertPlan("SELECT * FROM orders WHERE orderkey IN (SELECT orderkey FROM lineitem WHERE linenumber % 4 = 0)",
+                anyNodesTree(
+                        filterNode("semi_join_result",
```

Comment:
if you make a spelling error in `semi_join_result` it won't cause error when matching in `semiJoinNode`. You might want to use special naming for symbols that shouldn't be matched later on, e.g. (`any_x`). 

You could also use special functions in expressions, (e.g. `match_symbol(semi_join_result)`) to make sure it actually matches to something